examples/facenet_adversarial_faces/facenet_fgsm.py

```diff
@@ -0,0 +1,131 @@
+import facenet
+
+import tensorflow as tf
+import numpy as np
+from cleverhans.model import Model
+from cleverhans.attacks import FastGradientMethod
+
+import set_loader
+
+
+class InceptionResnetV1Model(Model):
+    model_path = "models/facenet/20170512-110547/20170512-110547.pb"
+
+    def __init__(self):
+        super(InceptionResnetV1Model, self).__init__()
+
+        # Load Facenet CNN
+        facenet.load_model(self.model_path)
+        # Save input and output tensors references
+        graph = tf.get_default_graph()
+        self.face_input = graph.get_tensor_by_name("input:0")
+        self.embedding_output = graph.get_tensor_by_name("embeddings:0")
+
+    def convert_to_classifier(self):
+        # Create victim_embedding placeholder
+        self.victim_embedding_input = tf.placeholder(
+            tf.float32,
+            shape=(None, 128))
+
+        # Squared Euclidean Distance between embeddings
+        distance = tf.reduce_sum(
+            tf.square(self.embedding_output - self.victim_embedding_input),
+            axis=1)
+
+        # Convert distance to a softmax vector
+        # 0.99 out of 4 is the distance threshold for the Facenet CNN
+        threshold = 0.99
+        score = tf.where(
+            distance > threshold,
+            0.5 + ((distance - threshold) * 0.5) / (4.0 - threshold),
+            0.5 * distance / threshold)
+        reverse_score = 1.0 - score
+        self.softmax_output = tf.transpose(tf.stack([reverse_score, score]))
+
+        # Save softmax layer
+        self.layer_names = []
+        self.layers = []
+        self.layers.append(self.softmax_output)
+        self.layer_names.append('probs')
+
+    def fprop(self, x, set_ref=False):
+        return dict(zip(self.layer_names, self.layers))
+
+
+with tf.Graph().as_default():
+    with tf.Session() as sess:
+        # Load model
+        model = InceptionResnetV1Model()
+        # Convert to classifier
+        model.convert_to_classifier()
+
+        # Load pairs of faces and their labels in one-hot encoding
+        faces1, faces2, labels = set_loader.load_testset(1000)
+
+        # Create victims' embeddings using Facenet itself
+        graph = tf.get_default_graph()
+        phase_train_placeholder = graph.get_tensor_by_name("phase_train:0")
+        feed_dict = {model.face_input: faces2,
+                     phase_train_placeholder: False}
+        victims_embeddings = sess.run(
+            model.embedding_output, feed_dict=feed_dict)
+
+        # Define FGSM for the model
+        fgsm = FastGradientMethod(model)
+        fgsm_params = {'eps': 0.01,
+                       'clip_min': 0.,
+                       'clip_max': 1.}
+        adv_x = fgsm.generate(model.face_input, **fgsm_params)
+
+        # Run FGSM
+        adv = faces1
+        steps = 1
+        for i in range(steps):

```

Comment:
It looks like you are trying to do something like the BasicIterativeMethod class (modulo the clipping), you might find that it simplifies your code.

Reply:
(This is a suggestion)

Reply:
I used FGSM + loops instead of the BasicIterativeMethod class because I thought it was easier to understand for the article that I wrote.

Regarding the clipping, it wasn't necessary as I was using epsilon as epsilon divided by the number of steps. Anyways I changed the code a little bit in order to use alpha instead of epsilon.